tls/design considerations.md

```diff
@@ -0,0 +1,47 @@
+# Design considerations for the libp2p TLS Handshake
+
+## Requirements 
+
+There are two main requirements that prevent us from using the straightforward way to run a TLS handshake (which would be to simply use the host key to create a self-signed certificate).
+
+1. We want to use different key types: RSA, ECDSA, and Ed25519, Secp256k1 (and maybe more in the future?).
+2. We want to be able to send the key type along with the key (see https://github.com/libp2p/specs/issues/111).
+
+The first point is problematic in practice, because Go currently only supports RSA and ECDSA certificates. Support for Ed25519 was planned for Go 1.12, but was deferred recently, and the Go team is now evaluating interest in this in order to prioritze their work, so this might or might not happen in Go 1.13. I'm not aware of any plans for Secp256k1 at the moment.
```

Comment:
You should fork the required library and use Ed25519 using agl's libraries.  ECDSA works but why bother.  RSA should not even be considered here.

Reply:
Really, we might wan to rephrase this in terms of ecosystem support. That is, TLS is a rather complicated protocol and we don't want to force libp2p implementations to add support for new key types.

Reply:
> You should fork the required library and use Ed25519 using agl's libraries.

I considered that, but concluded that's it's not a good solution. It would mean pulling in a forked x509 package, and in order to use that one, we'd also have to fork tls just to change the import path.

Reply:
As a general remark, in my opinion whether or not a library exists in the ecosystem should only be a very minor consideration when writing specs. You want your specs to do *the right thing*, not the thing that can be implemented the most easily.

Reply:
Which is why I didn't answer your question "Any estimate how long would it take to get this fixed in rustls?" (https://github.com/libp2p/specs/pull/151#issuecomment-475864990)
To me how long something takes to implement is not relevant when discussing specs.

Reply:
Filo has a branch at https://github.com/golang/go/issues/25355   I'm surprised this takes google so long given agl's role in Ed25519 being accepted by standards bodies.